kube-core/src/version.rs

```diff
@@ -122,34 +108,69 @@ impl FromStr for Version {
-            Version::Stable(v) => VersionSortKey::Stable(*v),
-            Version::Beta(v, beta) => VersionSortKey::Beta(*v, *beta),
-            Version::Alpha(v, alpha) => VersionSortKey::Alpha(*v, *alpha),
-            Version::Nonconformant(nc) => VersionSortKey::Nonconformant(Reverse(nc)),
-        }
-    }
+enum Stability {
+    Nonconformant,
+    Alpha,
+    Beta,
+    Stable,
 }
 
-impl Ord for Version {
-    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
-        self.to_sort_key().cmp(&other.to_sort_key())
-    }
+/// See [`Version::latest_stable`]
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+struct LatestStable {
+    stability: Stability,
+    major: u32,
+    minor: Option<u32>,
+    nonconformant: Option<Reverse<String>>,
 }
 
-impl PartialOrd for Version {
-    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
-        Some(self.cmp(other))
+impl Version {
+    /// An [`Ord`] for `Version` that prefers stable versions over later prereleases
+    ///
+    /// Implements the [Kubernetes version priority order](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority)
+    ///
+    /// For example:
+    ///
+    /// ```
+    /// # use kube_core::Version;
+    /// assert!(Version::Stable(2).latest_stable() > Version::Stable(1).latest_stable());
+    /// assert!(Version::Stable(1).latest_stable() > Version::Beta(1, None).latest_stable());
+    /// assert!(Version::Stable(1).latest_stable() > Version::Beta(2, None).latest_stable());
+    /// assert!(Version::Stable(2).latest_stable() > Version::Alpha(1, Some(2)).latest_stable());
+    /// assert!(Version::Stable(1).latest_stable() > Version::Alpha(2, Some(2)).latest_stable());
+    /// assert!(Version::Beta(1, None).latest_stable() > Version::Nonconformant("ver3".into()).latest_stable());
+    /// ```
+    ///
+    /// Note that the type of release matters more than the version numbers:
+    /// `Stable(x)` > `Beta(y)` > `Alpha(z)` > `Nonconformant(w)` for all `x`,`y`,`z`,`w`
+    pub fn latest_stable(&self) -> impl Ord {
+        match self {
+            &Version::Stable(major) => LatestStable {
+                stability: Stability::Stable,
+                major,
+                minor: None,
+                nonconformant: None,
+            },
+            &Version::Beta(major, minor) => LatestStable {
+                stability: Stability::Beta,
+                major,
+                minor,
+                nonconformant: None,
+            },
+            &Self::Alpha(major, minor) => LatestStable {
+                stability: Stability::Alpha,
+                major,
+                minor,
+                nonconformant: None,
+            },
+            Self::Nonconformant(nonconformant) => LatestStable {
+                stability: Stability::Nonconformant,
+                major: 0,
+                minor: None,
+                nonconformant: Some(Reverse(nonconformant.clone())),
```

Comment:
Not a huge fan of taking ownership here, but this'd fall away if we moved out the `Nonconformant` case (`Result<T, E>: Ord where T: Ord, E: Ord`, so I don't think it'd be a huge ergonomic problem), or if `Version` borrowed the nonconformant string rather than owning it.

Reply:
We do need to sort the nonconformant cases though, so we can't really remove it.

Reply:
No, in that case you'd run something like `Version::parse(ver).map(Version::latest_stable)`, producing a `Result<LatestStable, String>` which sorts the same as the current `LatestStable`.

Reply:
Presumably this would be `Version::parse(ver).map(Version::priority)` producing `Result<Priority, String>` now. Wouldn't this force us to also export our sort key types?

Reply:
Not really, `Result<impl Ord, String>` is fine as long as you never try to write down the type. That said, exporting the sort key types might still be desirable.